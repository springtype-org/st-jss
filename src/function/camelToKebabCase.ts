export const camelToKebabCase = (str: string): string =>
  str
    .split(/(?=[A-Z])/)
    .filter((v) => !!v)
    .join('-')
    .toLowerCase();
