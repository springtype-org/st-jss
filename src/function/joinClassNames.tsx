export const joinClassNames = (classes: Array<string | number>) =>
  classes
    .map((c) => `${c}`.trim())
    .filter((c) => !!c)
    .join('-');
