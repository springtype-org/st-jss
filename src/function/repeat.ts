export const repeat = (str: string, count: number) => {
  const result: Array<string> = [];
  for (let i = 0; i < count; i++) {
    result.push(str);
  }
  return result.join('');
};
