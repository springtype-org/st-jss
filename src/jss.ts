import { FontFaceProperties, KeyFrameProperties } from 'springtype-types';
import { API } from './interface/API';
import { ClassProperty } from './interface/ClassProperties';
import { makeStyles } from './makeStyles';
import { makeFont } from './makeFont';
import { makeAnimation } from './makeAnimation';

export const jss: API = {
  defaultUnit: 'px',
  classNamePrefix: 'st',
  instanceCounter: 0,
  makeAnimation: (animationName: string, style: Partial<KeyFrameProperties>) =>
    makeAnimation(animationName, style, jss),
  makeFont: (style: FontFaceProperties) => makeFont(style),
  makeStyles: <T extends string = string>(classProperty: ClassProperty<T>) => makeStyles<T>(classProperty, jss),
};
