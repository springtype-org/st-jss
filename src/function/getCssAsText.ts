import { camelToKebabCase } from './camelToKebabCase';
import { cssLengthRawKeys } from '../cssLengthRawKeys';

export const getCssAsText = (key: string, value: any, defaultUnit: string) => {
  const normalizedKey = camelToKebabCase(key);
  if (typeof value === 'number' && cssLengthRawKeys[normalizedKey]) {
    return `${normalizedKey}: ${value}${defaultUnit};`;
  }
  return `${normalizedKey}: ${value};`;
};
