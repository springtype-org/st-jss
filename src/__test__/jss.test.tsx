import { jss } from '../jss';
import { camelToKebabCase } from '../function/camelToKebabCase';
import { getCssAsText } from '../function/getCssAsText';
import { joinClassNames } from '../function/joinClassNames';

describe('jss', () => {
  it('is defined', () => {
    expect(jss).toBeDefined();
  });
});
describe('camelToKebabCase', () => {
  it('is equal', () => {
    expect(camelToKebabCase('backgroundColor')).toEqual('background-color');
    expect(camelToKebabCase('background-color')).toEqual('background-color');
  });
});
describe('getCssAsText', () => {
  it('is equal', () => {
    expect(getCssAsText('content', '""', 'px')).toEqual('content: "";');
    expect(getCssAsText('borderColor', '#fff', 'px')).toEqual('border-color: #fff;');
    expect(getCssAsText('height', '100%', 'px')).toEqual('height: 100%;');
    expect(getCssAsText('height', 1, 'px')).toEqual('height: 1px;');
    expect(getCssAsText('margin-top', 20, 'px')).toEqual('margin-top: 20px;');
  });
});
describe('joinClassNames', () => {
  it('is equal', () => {
    expect(joinClassNames(['st', 'header', 3])).toEqual('st-header-3');
    expect(joinClassNames(['st', 'container', 0])).toEqual('st-container-0');
  });
});

describe('joinClassNames', () => {
  it('check style output', () => {
    const [animationName, animationStyle] = jss.makeAnimation('animation', {
      '0%': { top: 0, left: 0, background: 'red' },
      '25%': { top: 0, left: 100, background: 'blue' },
      '50%': { top: 100, left: 100, background: 'yellow' },
      '75%': { top: 100, left: 0, background: 'green' },
      '100%': { top: 0, left: 0, background: 'red' },
    });
    const [classes, style] = jss.makeStyles({
      container: {
        // Normal jss
        fontFamily: 'Nonito',
        '@media (min-device-width: 800px)': {
          paddingTop: 50,
          flex: 1,
          ':hover': {
            backgroundColor: 'red',
          },
          ' span': {
            color: 'black',
            height: 4,
            fontSize: 20,
          },
          '.libraryClass': {
            color: 'black',
            height: 4,
            fontSize: 20,
          },
        },
      },
      animation: {
        paddingTop: 50,
        animationName,
        animationIterationCount: 'infinite',
        animationDuration: '4s',
      },
    });
    expect(classes).toMatchSnapshot();
    const { children: childrenOfStyle, ...otherPropsOfStyle } = style;
    const { children: childrenOfAnimation, ...otherPropsOfAnimation } = animationStyle;
    expect(childrenOfStyle[0]).toMatchSnapshot();
    expect(otherPropsOfStyle).toMatchSnapshot();
    expect(childrenOfAnimation[0]).toMatchSnapshot();
    expect(otherPropsOfAnimation).toMatchSnapshot();
  });
});
