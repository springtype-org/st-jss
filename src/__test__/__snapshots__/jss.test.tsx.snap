// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`joinClassNames check style output 1`] = `
Object {
  "animation": "st-animation-3",
  "container": "st-container-2",
}
`;

exports[`joinClassNames check style output 2`] = `
".st-container-2 {
  font-family: Nonito;
}
@media (min-device-width: 800px) {
  .st-container-2 {
    padding-top: 50px;
    flex: 1;
  }
  .st-container-2:hover {
    background-color: red;
  }
  .st-container-2 span {
    color: black;
    height: 4px;
    font-size: 20;
  }
  .st-container-2.libraryClass {
    color: black;
    height: 4px;
    font-size: 20;
  }
}
.st-animation-3 {
  padding-top: 50px;
  animation-name: st-animation-1;
  animation-iteration-count: infinite;
  animation-duration: 4s;
}"
`;

exports[`joinClassNames check style output 3`] = `
Object {
  "attributes": Object {
    "type": "text/css",
  },
  "type": "style",
}
`;

exports[`joinClassNames check style output 4`] = `
"@keyframes st-animation-1 {
  0% {
    top: 0px;
    left: 0px;
    background: red;
  }
  25% {
    top: 0px;
    left: 100px;
    background: blue;
  }
  50% {
    top: 100px;
    left: 100px;
    background: yellow;
  }
  75% {
    top: 100px;
    left: 0px;
    background: green;
  }
  100% {
    top: 0px;
    left: 0px;
    background: red;
  }
}"
`;

exports[`joinClassNames check style output 5`] = `
Object {
  "attributes": Object {
    "type": "text/css",
  },
  "type": "style",
}
`;
